pkg/datapath/iptables/iptables.go

```diff
@@ -1626,5 +1627,12 @@ func (m *IptablesManager) addCiliumENIRules() error {
 		"-A", ciliumPreMangleChain,
 		"-i", "lxc+",
 		"-m", "comment", "--comment", "cilium: primary ENI",
+		// Don't restore the mark when we are carrying the identity with mark,
+		// because we are using 7th bit to carry the uppermost bit of the identity
+		// and it is overlapping with ENI's mark. Below --restore-mark "restores"
```

Comment:
I understand why we want to do that, but why is it safe to do that?

It is because, if we are carrying the source identity in the packet mark, then we can assume the packet isn't a reply to a request from the external network? Are we sure about that assumption? Do we have anything in CI that ensures the PBR rule still has the intended effect?

Reply:
When I wrote my original patch, I was only considering the case of Pod->Pod traffic in the same node (local deliver case), and this fixes that case. Now I noticed that this would break the multi-node NodePort + multiple ENI + ClusterID <= 128 case (e.g., when the number of Pod on the node reaches the limit of AWS VPC CNI).

With direct routing mode, the Pod -> Another node traffic (e.g., NodePort reply) always has a `MARK_MAGIC_IDENTITY` set. Then it will bypass the connmark restore. When ClusterID > 128 and the Node only has a primary ENI attached, this will not become a problem since the packet anyways routed to the primary ENI.

However, with secondary ENI, this will become a problem, and the problem depends on the ClusterID. When ClusterID <= 128, the 7th bit is 0, so the NodePort reply will be routed to the secondary ENI. When ClusterID > 128, the 7th bit is 1, so the non-NodePort traffic will be routed to the primary ENI.

Reply:
Wrt the CI, we at least covered the case the ClusterID == 0 in `Conformance EKS` but not with the ClusterID > 128 and secondary ENI allocated.

Reply:
I tried the above scenario in the test PR and confirmed it failed.

PR (diff): https://github.com/cilium/cilium/pull/23279/files
Failed CI: https://github.com/cilium/cilium/actions/runs/4003372508/jobs/6871421429